File: RootToLeafSum.java
/**
Time Complexity : O(N) , N = number of elements
Space Complexity : O(D), D = depth of the tree
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
*/
class Solution 
{
    List<List<Integer>> output;
    int targetSum;
    
    public List<List<Integer>> pathSum(TreeNode root, int sum) 
    {
        output = new ArrayList<>();
        targetSum = sum;
        
        traverse(root, 0, new ArrayList<>());
        
        return output;
    }
    
    private void traverse(TreeNode root, int currentSum, List<Integer> currentList)
    {
       if(root == null)
           return;
        
        currentSum = currentSum + root.val;
        currentList.add(root.val);
        
        if(currentSum == targetSum)
        {
            if(root.left == null && root.right == null)
            {
                List<Integer> list = new ArrayList<>(currentList);
                output.add(list);
            }
        }
        
        if(root.left !=null)
            traverse(root.left, currentSum, currentList);
        
        if(root.right != null)
            traverse(root.right, currentSum, currentList);
        
        int currentListSize = currentList.size();
        currentList.remove(currentListSize-1);
    }
}

File: root_to_leaf_sum.py
# // Time Complexity : O(n^2)
# // Space Complexity : 0(n) 
# // Did this code successfully run on Leetcode : YES
# // Any problem you faced while coding this : NO

class Solution:
    
    def recurseTree(self, node, remainingSum, pathNodes, pathsList):
        
        if not node:
            return 
        
        # Add the current node to the path's list
        pathNodes.append(node.val)
        
        # Check if the current node is a leaf and also, if it
        # equals our remaining sum. If it does, we add the path to
        # our list of paths
        if remainingSum == node.val and not node.left and not node.right:
            pathsList.append(list(pathNodes))
        else:    
            # Else, we will recurse on the left and the right children
            self.recurseTree(node.left, remainingSum - node.val, pathNodes, pathsList)
            self.recurseTree(node.right, remainingSum - node.val, pathNodes, pathsList)
            
        # We need to pop the node once we are done processing ALL of it's
        # subtrees.
        pathNodes.pop()    
    
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        pathsList = []
        self.recurseTree(root, sum, [], pathsList)
        return pathsList